chainerx_cc/chainerx/routines/indexing.cc

```diff
@@ -199,35 +200,31 @@ Array Take(const Array& a, const Array& indices, int8_t axis) {
 }
 
 Array Where(const Array& condition, const Array& x, const Array& y) {
-    const Array& condition_cast = condition.dtype() != Dtype::kBool ? condition.AsType(Dtype::kBool) : condition;
-
-    if (x.dtype() != y.dtype()) {
-        throw DtypeError{"dtype of x and y should be same."};
-    }
+    Dtype out_dtype = ResultType(x, y);
+    const Array& x_cast = x.dtype() != out_dtype ? x.AsType(out_dtype) : x;
+    const Array& y_cast = y.dtype() != out_dtype ? y.AsType(out_dtype) : y;
```

Comment:
Could you move these casts to device implementation?

Reply:
Done. However, have kept 
`Dtype out_dtype = ResultType(x, y);` as is since it is consumed for while initializing `out`.